File: 28. Implement strStr().cpp
class Solution {
public:
    int strStr(string haystack, string needle) {
        int bound = haystack.size() - needle.size() + 1;
        for (int i = 0; i < bound; i++) {
            int j = 0;
            while (j < needle.size() && haystack[i + j] == needle[j])
                j++;
            if (j == needle.size())
                return i;
        }
        return -1;
    }
};

File: java/28. Implement strStr().java
// O(mn)
class Solution {
    public int strStr(String haystack, String needle) {
        if (needle.length() == 0) {
            return 0;
        } else if (haystack.length() == 0 || haystack.length() < needle.length()) {
            return -1;
        }
        for (int i = 0; i <= haystack.length() - needle.length(); ++i) {
            if (haystack.charAt(i) == needle.charAt(0)) {
                int j = 1;
                while (j < needle.length()) {
                    if (haystack.charAt(i + j) != needle.charAt(j)) {
                        break;
                    }
                    ++j;
                }
                if (j == needle.length()) {
                    return i;
                }
            }
        }
        return -1;
    }
}
